File: algorithms/easy/maximum_enemy_forts_that_can_be_captured.ex
defmodule Solution do
  @spec capture_forts(forts :: [integer]) :: integer
  def capture_forts(forts) do
    forts
    |> Enum.with_index()
    |> Enum.reduce({0, 0, 0}, fn {n, i}, {j, prev, ans} ->
      cond do
        n == 0 -> {j, prev, ans}
        n == -prev -> {i, n, max(ans, i - j - 1)}
        true -> {i, n, ans}
      end
    end)
    |> elem(2)
  end
end

File: algorithms/easy/maximum_enemy_forts_that_can_be_captured.cpp
class Solution {
public:
    int captureForts(vector<int>& forts) {
      int prev{0};
      int ans{0};
      for (int i = 0; i < forts.size(); i++) {
        if (forts[i] == 0) continue;
        if (forts[i] == -forts[prev]) {
          ans = max(ans, i - prev - 1);
        }
        prev = i;
      }
      return ans;
    }
};
